### PR TITLE
feat: Get projects in given folders

### PR DESCRIPTION
**Pull request by disq**
Introduces `folder_ids` and `folders_max_depth` to config. Deprecates `project_filter` (but it's still available if folders functionality is not used)
`folder_ids`, their children and `project_ids` are all used to build a list of projects.
Organization IDs can be specified as folders, when specified it will fetch all folders under that org.

Example config addition:
```hcl
    // Optional. List of folders to get projects from. Required permission: resourcemanager.projects.list
    // folder_ids = [ "organizations/<ORG_ID>", "folders/<FOLDER_ID>" ]
    // Optional. Maximum level of folders to recurse into
    // folders_max_depth = 5
```

Implements #184 

V1 [project list](https://cloud.google.com/resource-manager/reference/rest/v1/projects/list), used when no folders are specified to get a list of project IDs to run on:
> This method returns Projects in an unspecified order. This method is eventually consistent with project mutations; this means that a newly created project may not appear in the results or recent updates to an existing project may not be reflected in the results.

V3 [project list](https://cloud.google.com/resource-manager/reference/rest/v3/projects/list), used with folders, to query which projects reside in the given (or recursed) folders:
> Lists projects that are direct children of the specified folder or organization resource. list() provides a strongly consistent view of the projects underneath the specified parent resource. list() returns projects sorted based upon the (ascending) lexical ordering of their displayName. The caller must have resourcemanager.projects.list permission on the identified parent.

### CODE REVIEW

File: client/client.go [hunk 1]
```diff
@@ -85,14 +85,39 @@ func Configure(logger hclog.Logger, config interface{}) (schema.ClientMeta, erro
 		options = append(options, option.WithCredentialsJSON(serviceAccountKeyJSON))
 	}
 
-	var err error
-	if len(providerConfig.ProjectIDs) == 0 {
```
Comment (roneli) [inline client/client.go:89]: I understand you deprecated projectFilter but not ProjectIds?

Why not support the previous filter as well but make it mutually exclusive with Folder arguments?

i'll let @bbernays chime his thoughts here as well.

File: client/client.go [hunk 2]
```diff
@@ -85,14 +85,39 @@ func Configure(logger hclog.Logger, config interface{}) (schema.ClientMeta, erro
...
-		projects, err = getProjects(logger, providerConfig.ProjectFilter, options...)
+	if providerConfig.ProjectFilter != "" {
+		return nil, errors.New("ProjectFilter config option is deprecated")
+	}
+
+	if len(providerConfig.Folders) > 0 {
+		logger.Debug("Adding projects in specified folders")
```
Comment (roneli) [inline client/client.go:93]: nit: add folders to the logger

File: client/client.go [hunk 3]
```diff
@@ -85,14 +85,39 @@ func Configure(logger hclog.Logger, config interface{}) (schema.ClientMeta, erro
...
+		proj, err := getProjects(logger, providerConfig.Folders, options...)
+		if err != nil {
+			return nil, err
+		}
+		appendWithoutDupes(&projects, proj)
+	}
+	if providerConfig.FolderQuery != "" {
+		logger.Debug("Querying folders")
```
Comment (roneli) [inline client/client.go:101]: nit: add filter query

File: client/client.go [hunk 4]
```diff
@@ -123,24 +152,69 @@ func getProjects(logger hclog.Logger, filter string, options ...option.ClientOpt
...
+				return nil, err
+			}
+			for _, project := range output.Projects {
+				if project.State == "ACTIVE" {
+					projects = append(projects, project.ProjectId)
+				} else {
+					logger.Info("Project state is not active. Project will be ignored", "project_id", project.ProjectId, "project_state", project.State)
+					inactive++
+				}
+			}
```
Comment (roneli) [inline client/client.go:178]: maybe move this after the call or combine with the loop that removes dupes

**Review by roneli — COMMENTED** [review]
Added some comments.

File: client/client.go (hunk 1 above)
Comment (bbernays) [reply client/client.go:89]: We should keep the `projectIds` field for 2 reasons:
1. Backwards compatibility 
2. Not everyone uses Folders

File: client/client.go (hunk 4 above)
Comment (disq) [reply client/client.go:178]: I don't understand, maybe you parsed it wrong? `projects` is a slice of project ids, not project objects.

File: client/client.go (hunk 1 above)
Comment (disq) [reply client/client.go:89]: previous filter is not there in the `v3` api, you'd have to use the `Projects.Search` call (instead of `Projects.List`)
I can support it and keep appending all projects, but is it used?

**Review by roneli — APPROVED** [review]
LGTM, a minor nit would be for readability is maybe split the configure function into sub functions to find it easier to read.

**Comment by disq** [comment]
> LGTM, a minor nit would be for readability is maybe split the configure function into sub functions to find it easier to read.

I think it's already in sub-sections at least (big _if_ with folderID logic inside) so it's not that hard to follow.

**Review by yevgenypats — APPROVED** [review]

**Comment by yevgenypats** [comment]
Only thing is that now our tests are failing

**Comment by disq** [comment]
> Only thing is that now our tests are failing

It's not supposed to list folders if no `folder_ids` are specified in the config, so not sure what happened there

**Review by bbernays — APPROVED** [review]
